metapool/abs_quant_sample_info.py

```diff
@@ -0,0 +1,334 @@
+    MASS_STORAGE_TUBE_AND_STORAGE_LIQUID_AFTER_SAMPLE_MG_KEY,
+    STORAGE_LIQUID_LOT_NUM_STR_KEY]
+
+SLURRY_CALC_COLUMNS = INPUT_COLUMNS + CONFIG_COLUMNS + CORE_COLUMNS
+
+
+def add_abs_quant_metadata(
+        input_df: pandas.DataFrame, sample_density_g_ml_key: str,
+        storage_liquid_type: str, config_fp: Optional[str] = None) \
+        -> pandas.DataFrame:
+
+    """
+    Add metadata for absolute quantification calculations to the dataframe.
+
+    Parameters
+    ----------
+    input_df: pandas.DataFrame
+        Dataframe containing at least all the columns in CORE_COLUMNS.
+    sample_density_g_ml_key: str
+        Dictionary key for the density of the sample type for the samples in
+        the input df (usually stool).
+    storage_liquid_type: str
+        Type of storage liquid used with the samples in input df.
+    config_fp: str
+        Filepath to the yaml configuration file.  If None, the default
+        configuration file is used.
+
+    Returns
+    -------
+
+    calc_df: pandas.DataFrame
+        Extension of input calc_df with extensive additional columns for
+        absolute quantification calculations.
+    """
+
+    _validate_input_df(input_df)
+    calc_df = input_df.copy()
+
+    # ensure that the lot "number" (really lot ID) is stored a string
+    calc_df[STORAGE_LIQUID_LOT_NUM_STR_KEY] = calc_df[
+        STORAGE_LIQUID_LOT_NUM_STR_KEY].astype(str)
+
+    config_dict = _read_config_metadata(config_fp)
+    calc_df = _add_config_metadata(
+        calc_df, config_dict, sample_density_g_ml_key,
+        storage_liquid_type)
+
+    calc_df = _calc_abs_quant_metadata(calc_df)
+    return calc_df
+
+
+def _validate_input_df(input_df: pandas.DataFrame):
+    """
+    Validate that the input dataframe contains the required columns.
+
+    Parameters
+    ----------
+    input_df: pandas.DataFrame
+        Dataframe containing at least all the columns in CORE_COLUMNS.
+
+    Raises
+    -------
+    ValueError
+        If the input_df doesn't contain all the columns in CORE_COLUMNS.
+    """
+
+    input_cols = set(input_df.columns.tolist())
+    core_cols = set(CORE_COLUMNS)
+    if not core_cols.issubset(input_cols):
+        raise ValueError(
+            f"The input_df is missing the following required column(s): "
+            f"{', '.join(core_cols - input_cols)}")
+
+
+def _read_config_metadata(config_fp: str = None) -> dict:
+    """
+    Read the yaml configuration file into a dictionary.
+
+    Parameters
+    ----------
+    config_fp: str
+        Filepath to the yaml configuration file.  If None, the default
+        configuration file is used.
+
+    Returns
+    -------
+    config_dict: dict
+        Dictionary containing the configuration metadata from the config file.
+
+    """
+    if config_fp is None:
+        curr_dir = os.path.dirname(os.path.abspath(__file__))
+        config_fp = os.path.join(curr_dir, "data", DEFAULT_CONFIG_FNAME)
```

Comment:
It might be more consistent with the project to make DEFAULT_CONFIG_FNAME be DEFAULT_CONFIG_PATH and have it be the path to the file, rather than just the name. During deployment config files are often placed in a central location spanning multiple modules for convenience. IMHO an explicit exists() test here or earlier in the workflow might be desirable? It's going to raise an Error if it fails anyway.

Reply:
I have changed the default file name to a default file path.  I do not think an exists test is necessary because the basic error thrown if the path doesn't exist is quite informative on its own:

<img width="586" alt="Screenshot 2023-11-16 at 11 59 10 AM" src="https://github.com/biocore/metagenomics_pooling_notebook/assets/10677935/11f11b0d-3509-4b8b-8814-01ba7a9fe715">